_posts/2016-02-24-session-user-flow.markdown

```diff
@@ -0,0 +1,64 @@
+---
+layout: post
+title: "Use session variables to optimize your user flow"
+date: 2016-02-24 13:37:00
+categories: ["ruby", "rails", "sessions"]
+author: "schmierkov"
+---
+
+Sessions provide you a nice little data storage feature where the application does not need to get the informations directly from the database. So you do not have to persist data in your database and can easily store info to a user on the fly. This is a nice way to enhance the user experience on your page.
+
+Let's say that you want to show some users a new fancy sign up form and the rest the old form. If you store the version of the sign up form in a session variable, you don't need to persist this info in your database.
+
+<!--more-->
+
+That can be pretty handy in many ways, if you think about changing the content of a website based on the session's information. Handling a lot of session data can be pretty confusing, since the session is just a simple storage. That basically means that you have to organize your way of working with sessions. I once had the idea to put all this information into an object. This way I could test it and do wild things with the data, but the frustration comes as soon as you change your Object stored in your session.
+
+Basically this is what you should **NOT** do:
+
+```ruby
+class Location
+  def initialize(ip)
+    @location = location_request(ip)  
+  end
+
+  def city
+    @location.city
+  end
+  ...
+end
+class ApplicationController < ActionController::Base
+  before_action :set_location
+
+  def set_location(user_ip)
+    session[:user_location] = Location.new(user_ip)
+  end
+end
+```
+
+The problem is obvious: The app has no knowledge of the previously stored object and will raise an exception. A better way is to think about how and what you want to store in the session. Maybe a specific set of helper or controller methods just for handling the user navigation is enough. You can guide or force the user to a specific area of your website using the stored session information.
```

Comment:
@etagwerker ok, maybe it's not so obvious :( the problem that I'm trying to describe is that you should not store a ruby class in the session. the next time you rename or even add a method to this class, the user gets an exception.

I'm trying to describe a way, how to deal best with a lot of session variables by using helpers etc.. Am I making something more complicated?